### ISSUE TITLE
enable ghc-9.4 builds

### ISSUE THREAD

**Issue by shayne-fletcher** [ndmitchell/hlint#1413]
for some context on what remains to get ghc-9.4-* tested builds in CI, see https://github.com/ndmitchell/hlint/pull/1410#issuecomment-1213320218

the ghc-9.4.1 build itself is locally testable:
```
[2 of 2] Linking /Users/shayne/tmp/ghc-lib/9.4.1.20220813/ghc-9.4.1/dist-newstyle/build/x86_64-osx/ghc-9.4.1/hlint-9.4.1.20220813/x/hlint/build/hlint/hlint [Objects changed]
Testing (with refactoring)
Source annotations ...................
Input/outputs .............................................................................................................................................................
Hint names .......
Hint annotations .......

Tests passed (914)
+ exit 0
```

**Comment by ndmitchell** [comment]
Seems like we are waiting on ghc-exactprint to release?

**Comment by shayne-fletcher** [comment]
> Seems like we are waiting on ghc-exactprint to release?

it wouldn't hurt to have one. it's blocking ghc-9.4.1 builds in CI

**Comment by LeventErkok** [comment]
Just adding the failures I see when trying to build with GHC 9.4.2:

```
[ 4 of 60] Compiling GHC.Util.ApiAnnotation ( src/GHC/Util/ApiAnnotation.hs, dist/build/GHC/Util/ApiAnnotation.o, dist/build/GHC/Util/ApiAnnotation.dyn_o )

src/GHC/Util/ApiAnnotation.hs:39:28: error:
    Not in scope: data constructor ‘EpaDocCommentNamed’
    Suggested fix:
      Perhaps use ‘EpaDocComment’ (imported from GHC.Parser.Annotation)
   |
39 | comment_ (L _ (EpaComment (EpaDocCommentNamed s) _)) = s
   |                            ^^^^^^^^^^^^^^^^^^

src/GHC/Util/ApiAnnotation.hs:40:28: error:
    Not in scope: data constructor ‘EpaDocCommentPrev’
    Suggested fix:
      Perhaps use ‘EpaDocComment’ (imported from GHC.Parser.Annotation)
   |
40 | comment_ (L _ (EpaComment (EpaDocCommentPrev s) _)) = s
   |                            ^^^^^^^^^^^^^^^^^

src/GHC/Util/ApiAnnotation.hs:41:28: error:
    Not in scope: data constructor ‘EpaDocCommentNext’
    Suggested fix:
      Perhaps use ‘EpaDocComment’ (imported from GHC.Parser.Annotation)
   |
41 | comment_ (L _ (EpaComment (EpaDocCommentNext s) _)) = s
   |                            ^^^^^^^^^^^^^^^^^

src/GHC/Util/ApiAnnotation.hs:42:28: error:
    Not in scope: data constructor ‘EpaDocSection’
   |
42 | comment_ (L _ (EpaComment (EpaDocSection _ s) _)) = s
   |                            ^^^^^^^^^^^^^
[ 5 of 60] Compiling GHC.Util.Brackets ( src/GHC/Util/Brackets.hs, dist/build/GHC/Util/Brackets.o, dist/build/GHC/Util/Brackets.dyn_o )

src/GHC/Util/Brackets.hs:40:7: error:
    Not in scope: data constructor ‘HsRecFld’
   |
40 |       HsRecFld{} -> False
   |       ^^^^^^^^

src/GHC/Util/Brackets.hs:51:7: error:
    Not in scope: data constructor ‘HsBracket’
    Suggested fix: Perhaps use ‘HsBracketTc’ (imported from GHC.Hs)
   |
51 |       HsBracket{} -> True
   |       ^^^^^^^^^
[ 6 of 60] Compiling GHC.Util.DynFlags ( src/GHC/Util/DynFlags.hs, dist/build/GHC/Util/DynFlags.o, dist/build/GHC/Util/DynFlags.dyn_o )
[ 7 of 60] Compiling GHC.Util.FreeVars ( src/GHC/Util/FreeVars.hs, dist/build/GHC/Util/FreeVars.o, dist/build/GHC/Util/FreeVars.dyn_o )

src/GHC/Util/FreeVars.hs:112:18: error:
    Not in scope: data constructor ‘HsBracket’
    Suggested fix: Perhaps use ‘HsBracketTc’ (imported from GHC.Hs)
    |
112 |   freeVars (L _ (HsBracket _ (ExpBr _ e))) = freeVars e
    |                  ^^^^^^^^^

src/GHC/Util/FreeVars.hs:113:18: error:
    Not in scope: data constructor ‘HsBracket’
    Suggested fix: Perhaps use ‘HsBracketTc’ (imported from GHC.Hs)
    |
113 |   freeVars (L _ (HsBracket _ (VarBr _ _ v))) = Set.fromList [occName (unLoc v)]
    |                  ^^^^^^^^^

src/GHC/Util/FreeVars.hs:115:17: error:
    Not in scope: data constructor ‘HsConLikeOut’
    |
115 |   freeVars (L _ HsConLikeOut{}) = mempty -- After typechecker.
    |                 ^^^^^^^^^^^^

src/GHC/Util/FreeVars.hs:116:17: error:
    Not in scope: data constructor ‘HsRecFld’
    |
116 |   freeVars (L _ HsRecFld{}) = mempty -- Variable pointing to a record selector.
    |                 ^^^^^^^^

src/GHC/Util/FreeVars.hs:121:17: error:
    Not in scope: data constructor ‘HsRnBracketOut’
    |
121 |   freeVars (L _ HsRnBracketOut{}) = mempty -- Renamer produces these.
    |                 ^^^^^^^^^^^^^^

src/GHC/Util/FreeVars.hs:122:17: error:
    Not in scope: data constructor ‘HsTcBracketOut’
    |
122 |   freeVars (L _ HsTcBracketOut{}) = mempty -- Typechecker produces these.
    |                 ^^^^^^^^^^^^^^

src/GHC/Util/FreeVars.hs:159:21: error:
    Not in scope: data constructor ‘HsRecField’
    Suggested fix:
      Perhaps use one of these:
        ‘HsRecFields’ (imported from GHC.Hs),
        ‘HsGetField’ (imported from GHC.Hs),
        ‘HsRecSel’ (imported from GHC.Hs)
    |
159 |    freeVars o@(L _ (HsRecField _ x _ True)) = Set.singleton $ occName $ unLoc $ rdrNameFieldOcc $ unLoc x -- a pun
    |                     ^^^^^^^^^^

src/GHC/Util/FreeVars.hs:160:21: error:
    Not in scope: data constructor ‘HsRecField’
    Suggested fix:
      Perhaps use one of these:
        ‘HsRecFields’ (imported from GHC.Hs),
        ‘HsGetField’ (imported from GHC.Hs),
        ‘HsRecSel’ (imported from GHC.Hs)
    |
160 |    freeVars o@(L _ (HsRecField _ _ x _)) = freeVars x
    |                     ^^^^^^^^^^

src/GHC/Util/FreeVars.hs:162:30: error:
    Not in scope: type constructor or class ‘HsRecField'’
    Suggested fix:
      Perhaps use one of these:
        ‘HsRecField’ (imported from GHC.Hs),
        ‘HsRecFields’ (imported from GHC.Hs),
        ‘LHsRecField’ (imported from GHC.Hs)
    |
162 | instance FreeVars (LocatedA (HsRecField' (AmbiguousFieldOcc GhcPs) (LocatedA (HsExpr GhcPs)))) where
    |                              ^^^^^^^^^^^

src/GHC/Util/FreeVars.hs:163:18: error:
    Not in scope: data constructor ‘HsRecField’
    Suggested fix:
      Perhaps use one of these:
        ‘HsRecFields’ (imported from GHC.Hs),
        ‘HsGetField’ (imported from GHC.Hs),
        ‘HsRecSel’ (imported from GHC.Hs)
    |
163 |   freeVars (L _ (HsRecField _ _ x _)) = freeVars x
    |                  ^^^^^^^^^^

src/GHC/Util/FreeVars.hs:165:30: error:
    Not in scope: type constructor or class ‘HsRecField'’
    Suggested fix:
      Perhaps use one of these:
        ‘HsRecField’ (imported from GHC.Hs),
        ‘HsRecFields’ (imported from GHC.Hs),
        ‘LHsRecField’ (imported from GHC.Hs)
    |
165 | instance FreeVars (LocatedA (HsRecField' (FieldLabelStrings GhcPs) (LocatedA (HsExpr GhcPs)))) where
    |                              ^^^^^^^^^^^

src/GHC/Util/FreeVars.hs:166:18: error:
    Not in scope: data constructor ‘HsRecField’
    Suggested fix:
      Perhaps use one of these:
        ‘HsRecFields’ (imported from GHC.Hs),
        ‘HsGetField’ (imported from GHC.Hs),
        ‘HsRecSel’ (imported from GHC.Hs)
    |
166 |   freeVars (L _ (HsRecField _ _ x _)) = freeVars x
    |                  ^^^^^^^^^^

src/GHC/Util/FreeVars.hs:192:18: error:
    Not in scope: data constructor ‘HsRecField’
    Suggested fix:
      Perhaps use one of these:
        ‘HsRecFields’ (imported from GHC.Hs),
        ‘HsGetField’ (imported from GHC.Hs),
        ‘HsRecSel’ (imported from GHC.Hs)
    |
192 |    allVars (L _ (HsRecField _ _ x _)) = allVars x
-----------------------------------------------------------------------------
    |                  ^^^^^^^^^^
[ 8 of 60] Compiling GHC.Util.HsDecl  ( src/GHC/Util/HsDecl.hs, dist/build/GHC/Util/HsDecl.o, dist/build/GHC/Util/HsDecl.dyn_o )
[ 9 of 60] Compiling GHC.Util.Scope   ( src/GHC/Util/Scope.hs, dist/build/GHC/Util/Scope.o, dist/build/GHC/Util/Scope.dyn_o )

src/GHC/Util/Scope.hs:37:19: error:
    • Couldn't match type ‘GHC.Types.PkgQual.RawPkgQual’
                     with ‘Maybe StringLiteral’
      Expected: Maybe StringLiteral
        Actual: ImportDeclPkgQual GhcPs
    • In the expression: ideclPkgQual x
      In an equation for ‘pkg’: pkg (L _ x) = ideclPkgQual x
      In an equation for ‘scopeCreate’:
          scopeCreate xs
            = Scope $ [prelude | not $ any isPrelude res] ++ res
            where
                pkg :: LImportDecl GhcPs -> Maybe StringLiteral
                pkg (L _ x) = ideclPkgQual x
                res :: [LImportDecl GhcPs]
                res
                  = [x |
                       x <- hsmodImports xs,
                       pkg x /= Just (StringLiteral NoSourceText (fsLit "hint") Nothing)]
                ....
   |
37 |     pkg (L _ x) = ideclPkgQual x
   |                   ^^^^^^^^^^^^^^
[10 of 60] Compiling GHC.Util.SrcLoc  ( src/GHC/Util/SrcLoc.hs, dist/build/GHC/Util/SrcLoc.o, dist/build/GHC/Util/SrcLoc.dyn_o )

src/GHC/Util/SrcLoc.hs:21:47: error:
    • Couldn't match expected type: GHC.Data.Strict.Maybe BufSpan
                  with actual type: Maybe a0
      NB: ‘GHC.Data.Strict.Maybe’
            is defined in ‘GHC.Data.Strict’
                in package ‘ghc-lib-parser-9.4.2.20220822’
          ‘Maybe’ is defined in ‘GHC.Maybe’ in package ‘base-4.17.0.0’
    • In the second argument of ‘RealSrcSpan’, namely ‘Nothing’
      In the expression: RealSrcSpan (anchor (getLoc o)) Nothing
      In an equation for ‘getAncLoc’:
          getAncLoc o = RealSrcSpan (anchor (getLoc o)) Nothing
   |
21 | getAncLoc o = RealSrcSpan (anchor (getLoc o)) Nothing
   |                                               ^^^^^^^
```

**Comment by shayne-fletcher** [comment]
Hi @LeventErkok 

> Just adding the failures I see when trying to build with GHC 9.4.2:
> ...

(1) For the avoidance of doubt, there is no issue building hlint (HEAD) with ghc-9.4.2:
```
$ cd ~/project/hlint
$ cabal new-build
Resolving dependencies...
Build profile: -w ghc-9.4.2 -O1
...
[2 of 2] Linking /Users/shaynefletcher/project/hlint/dist-newstyle/build/x86_64-osx/ghc-9.4.2/hlint-3.4.1/x/hlint/build/hlint/hlint
```

(2) If we try to build hlint-3.4.1 (the most recently released) using ghc-9.4.2 we (correctly) fail to get a build-plan:
```
cd ~/tmp
cabal get hlint-3.4.1
cd hlint-3.4.1
cabal new-build
❯ cabal new-build
Resolving dependencies...
Error: cabal: Could not resolve dependencies:
[__0] trying: hlint-3.4.1 (user goal)
[__1] next goal: ghc-lib-parser (dependency of hlint +/-ghc-lib)
[__1] rejecting: ghc-lib-parser-9.4.2.20220822, ghc-lib-parser-9.4.1.20220807
(conflict: hlint +/-ghc-lib => ghc-lib-parser>=9.2 && <9.3)
[__1] trying: ghc-lib-parser-9.2.4.20220729
[__2] next goal: base (dependency of hlint)
[__2] rejecting: base-4.17.0.0/installed-4.17.0.0 (conflict: ghc-lib-parser =>
base>=4.14 && <4.17)
...
```
This is by design. Building a 9.2 parse tree ghc-lib-parser with a 9.4 compiler is not supported.

(3) The "errors" reported  I guess result from attempting to work-around the constraint  in (2) by doing e.g. `cabal new-build --allow-newer=ghc-lib-parser`:
```
src/GHC/Util/ApiAnnotation.hs:39:28: error:
    Not in scope: data constructor ‘EpaDocCommentNamed’
    Suggested fix:
      Perhaps use ‘EpaDocComment’ (imported from GHC.Parser.Annotation)
   |
39 | comment_ (L _ (EpaComment (EpaDocCommentNamed s) _)) = s
   |                            ^^^^^^^^^^^^^^^^^^
```
The attempted build-plan is invalid and this build failure is not indicative of an error.

**Comment by LeventErkok** [comment]
Thanks @shayne-fletcher

That's correct indeed. So, would be great to have a release of the HEAD to hackage? Is there some reason that can't happen?

**Comment by ndmitchell** [comment]
Hmm - what changed in HEAD that means it works? Since it isn't in the changelog... But I'll make a release shortly

**Comment by ndmitchell** [comment]
The comment from Levent suggests that things work better once we make a release of HEAD to Hackage. Either that's true (in which case something changed) or false (and it should already work)?

**Comment by shayne-fletcher** [comment]
the latter i think. nothing has changed particularly. since https://github.com/ndmitchell/hlint/pull/1410 we have been able to build with ghc-9.4.1 (& ghc-9.4.2) we just haven't been able to get CI green due to apply-refact.

**Comment by ndmitchell** [comment]
OK, so we were missing the GHC 9.4 parse tree upgrade in the changelog (my bad). We probably work with GHC 9.4, but the apply-refact won't work there yet. I'll push out a release.

**Comment by LeventErkok** [comment]
@ndmitchell Thanks! I can confirm that the new release works with GHC 9.4.2.
